riscv/src/continuations/bootloader.rs

```diff
@@ -337,11 +342,140 @@ final_memory_hash_ok:
 // Otherwise, we jump to the PC.
 jump_to_bootloader_input {PC_INDEX};
 
-end_of_bootloader:
+// END OF BOOTLOADER
+
 "#
     ));
 
-    bootloader.push_str("\n// END OF BOOTLOADER\n");
+    bootloader.push_str(&format!(
+        r#"
+// START OF SHUTDOWN ROUTINE
+//
+// This code is currently never executed in practice!
+//
+// The shutdown routine is responsible for:
+// - Validating that the final register values are equal to those in the bootloader inputs
+//   (which are exposed as public outputs)
+// - Validating that the final page hashes are equal to the claimed values provided in the
+//   bootloader inputs (which have previously been used to update the Merkle tree)
+//
+// During the execution of the shutdown routine, registers are used as follows:
+// - x1: Number of pages (constant throughout the execution)
+// - x2: Current page index
+// - x3: Current page number
+// - P0-P11: Hash registers, used to compute the page hash
+
+shutdown_start:
+
+// Assert final register values are as claimed
+// Note that we cannot assert that the final PC is correct, because it will already
+// have changed at this point. This will need to be done by whatever mechanism is used
+// to jump to the shutdown routine.
+"#
+    ));
+
+    // Go over all registers except the PC
+    let register_iter = REGISTER_NAMES.iter().take(REGISTER_NAMES.len() - 1);
+
+    for (i, reg) in register_iter.enumerate() {
+        let reg = reg.strip_prefix("main.").unwrap();
+        bootloader.push_str(&format!(
+            "assert_bootloader_input {}, {reg};\n",
+            i + REGISTER_NAMES.len()
+        ));
+    }
+
+    bootloader.push_str(&format!(
+        r#"
+// Number of pages
+x1 <== load_bootloader_input({NUM_PAGES_INDEX});
+x1 <== wrap(x1);
+
+// Current page index
+x2 <=X= 0;
+
+branch_if_zero x1, shutdown_end_page_loop;
+
+shutdown_start_page_loop:
+
+// Page number
+x3 <== load_bootloader_input(x2 * {BOOTLOADER_INPUTS_PER_PAGE} + {PAGE_INPUTS_OFFSET});
+x3 <== and(x3, {PAGE_NUMBER_MASK});
+
+// Store & hash {WORDS_PER_PAGE} page words. This is an unrolled loop that for each each word:
+// - Loads the word at the address x3 * {PAGE_SIZE_BYTES} + i * {BYTES_PER_WORD}
+//   into the P{{(i % 4) + 4}} register
+// - If i % 4 == 3: Hashes registers P0-P11, storing the result in P0-P3
+//
+// At the end of the loop, we'll have a linear hash of the page in P0-P3, using a Merkle-Damgård
+// construction. The initial P0-P3 values are 0, and the capacity (P8-P11) is 0 throughout the
+// execution of the shutdown routine.
+
+P0 <=X= 0;
+P1 <=X= 0;
+P2 <=X= 0;
+P3 <=X= 0;
+P4 <=X= 0;
+P5 <=X= 0;
+P6 <=X= 0;
+P7 <=X= 0;
+P8 <=X= 0;
+P9 <=X= 0;
+P10 <=X= 0;
+P11 <=X= 0;
+"#,
+    ));
+
+    for i in 0..WORDS_PER_PAGE {
+        let reg_index = (i % 4) + 4;
+        bootloader.push_str(&format!(
+            "P{reg_index}, x0 <== mload(x3 * {PAGE_SIZE_BYTES} + {i} * {BYTES_PER_WORD});\n"
```

Comment:
Note that while we need two instruction per word in the bootloader (one to load the bootloader input, one to write to memory), we only need 1 here!